### PR TITLE
[Internal] Query: Fixes ORDER BY issue when partial partition key is specified in RequestOptions in a query to sub-partitioned container

### LINKED ISSUE #4563: Query on one partition returns data from other partitions on multi-partition container
**Describe the bug**
When I run a targeted multi-partition query with a partition key that includes the first level of subpartitioning, the response contains items from different partitions. This happens only if QueryRequestOptions.PartitionKey is set.

**To Reproduce**
I created [a console application](https://github.com/user-attachments/files/16021167/GetItemsWrongPartitions.zip) to reproduce the issue. It is based on our test DB. Please download the solution, open the Program.cs and update key (can be supplied separately).

**Expected behavior**
The output should contains items from 06f43ec1-b920-440b-91fb-6dfc5b3af47d partition .

**Actual behavior**
The output contains items from different partitions.
PartitionKey: 344ea563-1653-408b-8fd0-7735e1ccd4bf; 
PartitionKey: adf1443d-8b8f-4f8e-9421-093420d9d290; 

Environment summary
SDK Version: 3.41.0
OS Version: Windows 11 Enterprise

### PR DESCRIPTION
**Pull request by adityasa**
# Pull Request Template

## Description

This changes addresses the issue that when an ORDER BY query is issued to a sub-partitioned container with partial partition key in the request options, documents outside the specified logical partition can also be retrieved. This happens if the logical partition range specified by partition key is split across multiple physical partitions (as with all HPK related issues so far) and one of the physical partitions contains data from other logical partitions (which is very likely, especially if a split has taken place).
With the previous fix made in following PR, which is still a required fix, we ensured that in cases when logical partition is split across multiple physical partitions, we perform a cross-partition ordering in the client:
https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4507

However, as a result of this change, information regarding the logical partition key is not honored and each physical partition gets queried fully. This is because enumerator is scoped to full partition and when request is issued, epk range header is not specified in the request (only partition id is specified)

This change refactors the existing logic (used in another pipeline) to limit the FeedRangeEpk used by the enumerator to the intersection of physical partition range and logical partition feed range. This effectively limits the query epk range (by including correct header) for cases when a physical partition contains data from other logical partitions as well.

As far as testing is concerned, the InMemoryContainer was already able to repro this case as part of previous change. And it was originally thought to be the issue with the test infra in the previous change and a fix was applied to suppress this issue. The change removes this suppression and it was confirmed that existing ORDER BY queries in SubPartitionedTests repro the issue.

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] This change requires a documentation update

## 

To automatically close an issue: closes #4563

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/HierarchicalPartitionUtil.cs [hunk 1]
```diff
@@ -0,0 +1,101 @@
+﻿// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  All rights reserved.
+// ------------------------------------------------------------
+
+namespace Microsoft.Azure.Cosmos.Query.Core.Pipeline.CrossPartition
+{
+    using System;
+    using Microsoft.Azure.Cosmos.Query.Core.QueryClient;
+
+    internal static class HierarchicalPartitionUtil
```
Comment (neildsh) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/HierarchicalPartitionUtil.cs:10]: >HierarchicalPartitionUtil [](http://example.com/codeflow?start=26&length=25)

[Nit] consider `HierarchicalPartitionKeyUtils`

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs [hunk 2]
```diff
@@ -90,7 +94,8 @@ private enum ExecutionState
             QueryExecutionOptions queryPaginationOptions,
             int maxConcurrency,
             bool nonStreamingOrderBy,
-            CosmosElement continuationToken)
+            CosmosElement continuationToken,
+            ContainerQueryProperties containerQueryProperties)
```
Comment (neildsh) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs:98]: [Nit] Please move to be the second parameter. Mandatory parameters should not be after optional parameters. Similarly, the parameters are ordered from most general to most specific.

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs [hunk 3]
```diff
@@ -146,14 +152,16 @@ private enum ExecutionState
...
+                maxConcurrency,
+                containerQueryProperties));
         }
 
         private static async ValueTask MoveNextAsync_InitializeAsync_HandleSplitAsync(
             IDocumentContainer documentContainer,
             Queue<(OrderByQueryPartitionRangePageAsyncEnumerator enumerator, OrderByContinuationToken token)> uninitializedEnumeratorsAndTokens,
             OrderByQueryPartitionRangePageAsyncEnumerator uninitializedEnumerator,
             OrderByContinuationToken token,
+            ContainerQueryProperties containerQueryProperties,
```
Comment (neildsh) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs:164]: [Nit] Similar to above: please move to be the second parameter

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs [hunk 4]
```diff
@@ -298,6 +308,7 @@ private sealed class StreamingOrderByCrossPartitionQueryPipelineStage : IQueryPi
             private readonly Queue<(OrderByQueryPartitionRangePageAsyncEnumerator enumerator, OrderByContinuationToken token)> uninitializedEnumeratorsAndTokens;
             private readonly QueryExecutionOptions queryPaginationOptions;
             private readonly int maxConcurrency;
+            private readonly ContainerQueryProperties containerQueryProperties;
```
Comment (neildsh) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs:311]: [Nit] Similar to above: please move to be the second member

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs [hunk 5]
```diff
@@ -319,7 +330,8 @@ private static class Expressions
                 QueryExecutionOptions queryPaginationOptions,
                 int maxConcurrency,
                 IEnumerable<(OrderByQueryPartitionRangePageAsyncEnumerator, OrderByContinuationToken)> uninitializedEnumeratorsAndTokens,
-                QueryState state)
+                QueryState state,
+                ContainerQueryProperties containerQueryProperties)
```
Comment (neildsh) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs:334]: [Nit] Similar to above: please move to be the second parameter

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs [hunk 6]
```diff
@@ -328,6 +340,7 @@ private static class Expressions
                 this.maxConcurrency = maxConcurrency < 0 ? throw new ArgumentOutOfRangeException($"{nameof(maxConcurrency)} must be a non negative number.") : maxConcurrency;
                 this.uninitializedEnumeratorsAndTokens = new Queue<(OrderByQueryPartitionRangePageAsyncEnumerator, OrderByContinuationToken)>(uninitializedEnumeratorsAndTokens ?? throw new ArgumentNullException(nameof(uninitializedEnumeratorsAndTokens)));
                 this.state = state ?? InitializingQueryState;
+                this.containerQueryProperties = containerQueryProperties;
```
Comment (neildsh) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs:343]: Please add null check

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs [hunk 7]
```diff
@@ -336,7 +349,8 @@ private static class Expressions
                 PriorityQueue<OrderByQueryPartitionRangePageAsyncEnumerator> enumerators,
                 Queue<(OrderByQueryPartitionRangePageAsyncEnumerator enumerator, OrderByContinuationToken token)> uninitializedEnumeratorsAndTokens,
                 QueryExecutionOptions queryPaginationOptions,
-                int maxConcurrency)
+                int maxConcurrency,
+                ContainerQueryProperties containerQueryProperties)
```
Comment (neildsh) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs:353]: [Nit] Similar to above: please move to be the second parameter

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs [hunk 8]
```diff
@@ -345,6 +359,7 @@ private static class Expressions
                 this.queryPaginationOptions = queryPaginationOptions ?? throw new ArgumentNullException(nameof(queryPaginationOptions));
                 this.maxConcurrency = maxConcurrency;
                 this.state = InitializingQueryState;
+                this.containerQueryProperties = containerQueryProperties;
```
Comment (neildsh) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs:362]: Please add null check

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs [hunk 9]
```diff
@@ -798,7 +816,8 @@ public ValueTask<bool> MoveNextAsync(ITrace trace, CancellationToken cancellatio
                 IReadOnlyList<OrderByColumn> orderByColumns,
                 QueryExecutionOptions queryPaginationOptions,
                 int maxConcurrency,
-                CosmosElement continuationToken)
+                CosmosElement continuationToken,
+                ContainerQueryProperties containerQueryProperties)
```
Comment (neildsh) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs:820]: [Nit] Similar to above: please move to be the second parameter

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs [hunk 10]
```diff
@@ -779,15 +795,17 @@ public ValueTask<bool> MoveNextAsync(ITrace trace, CancellationToken cancellatio
                 PriorityQueue<OrderByQueryPartitionRangePageAsyncEnumerator> enumerators,
                 Queue<(OrderByQueryPartitionRangePageAsyncEnumerator enumerator, OrderByContinuationToken token)> uninitializedEnumeratorsAndTokens,
                 QueryExecutionOptions queryPaginationOptions,
-                int maxConcurrency)
+                int maxConcurrency,
+                ContainerQueryProperties containerQueryProperties)
```
Comment (neildsh) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs:799]: [Nit] Similar to above: please move to be the second parameter

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs [hunk 11]
```diff
@@ -1768,7 +1794,8 @@ private async Task<BufferedOrderByResults> MoveNextAsync_InitializeAsync(ITrace
                 Cosmos.PartitionKey? partitionKey,
                 IReadOnlyList<OrderByColumn> orderByColumns,
                 QueryExecutionOptions queryPaginationOptions,
-                int maxConcurrency)
+                int maxConcurrency,
+                ContainerQueryProperties containerQueryProperties)
```
Comment (neildsh) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs:1798]: [Nit] Similar to above: please move to be the second parameter

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs [hunk 12]
```diff
@@ -1795,14 +1823,18 @@ private sealed class OrderByCrossPartitionRangePageEnumerator : ITracingAsyncEnu
 
             private readonly Queue<(OrderByQueryPartitionRangePageAsyncEnumerator enumerator, OrderByContinuationToken token)> enumeratorsAndTokens;
 
+            private readonly ContainerQueryProperties containerQueryProperties;
```
Comment (neildsh) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs:1826]: [Nit] Similar to above: please move to be the second member

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs [hunk 13]
```diff
@@ -1795,14 +1823,18 @@ private sealed class OrderByCrossPartitionRangePageEnumerator : ITracingAsyncEnu
...
 
             private OrderByCrossPartitionRangePageEnumerator(
                 IDocumentContainer documentContainer,
-                Queue<(OrderByQueryPartitionRangePageAsyncEnumerator enumerator, OrderByContinuationToken token)> enumeratorsAndTokens)
+                Queue<(OrderByQueryPartitionRangePageAsyncEnumerator enumerator, OrderByContinuationToken token)> enumeratorsAndTokens,
+                ContainerQueryProperties containerQueryProperties)
             {
                 this.documentContainer = documentContainer ?? throw new ArgumentNullException(nameof(documentContainer));
                 this.enumeratorsAndTokens = enumeratorsAndTokens ?? throw new ArgumentNullException(nameof(enumeratorsAndTokens));
+                this.containerQueryProperties = containerQueryProperties;
```
Comment (neildsh) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs:1837]: Please add null check

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs [hunk 14]
```diff
@@ -1812,6 +1844,7 @@ private sealed class OrderByCrossPartitionRangePageEnumerator : ITracingAsyncEnu
                 Cosmos.PartitionKey? partitionKey,
                 QueryExecutionOptions queryPaginationOptions,
                 int maxConcurrency,
+                ContainerQueryProperties containerQueryProperties,
```
Comment (neildsh) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs:1847]: [Nit] Similar to above: please move to be the second parameter

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByQueryPartitionRangePageAsyncEnumerator.cs [hunk 15]
```diff
@@ -25,15 +26,17 @@ internal sealed class OrderByQueryPartitionRangePageAsyncEnumerator : PartitionR
             PartitionKey? partitionKey,
             QueryExecutionOptions queryPaginationOptions,
             string filter,
-            PrefetchPolicy prefetchPolicy)
+            PrefetchPolicy prefetchPolicy,
+            ContainerQueryProperties containerQueryProperties)
```
Comment (neildsh) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByQueryPartitionRangePageAsyncEnumerator.cs:30]: [Nit] Similar to above: please move to be the second parameter

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByQueryPartitionRangePageAsyncEnumerator.cs [hunk 16]
```diff
@@ -105,22 +108,25 @@ public OrderByQueryPartitionRangePageAsyncEnumerator CloneAsFullyBufferedEnumera
...
 
             public InnerEnumerator(
                 IQueryDataSource queryDataSource,
                 SqlQuerySpec sqlQuerySpec,
                 FeedRangeState<QueryState> feedRangeState,
                 PartitionKey? partitionKey,
                 QueryExecutionOptions queryPaginationOptions,
-                string filter)
+                string filter,
+                ContainerQueryProperties containerQueryProperties)
```
Comment (neildsh) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByQueryPartitionRangePageAsyncEnumerator.cs:121]: [Nit] Similar to above: please move to be the second parameter

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Pagination/InMemoryContainer.cs [hunk 17]
```diff
@@ -515,7 +515,7 @@ static Task<TryCatch<Record>> CreateNotFoundException(CosmosElement partitionKey
                 }
 
                 List<CosmosObject> documents = new List<CosmosObject>();
-                foreach (Record record in records.Where(r => IsRecordWithinFeedRange(r, feedRangeState.FeedRange, this.partitionKeyDefinition) && IsRecordWithinQueryPartition(r, this.queryRequestOptions, this.partitionKeyDefinition)))
+                foreach (Record record in records.Where(r => IsRecordWithinFeedRange(r, feedRangeState.FeedRange, this.partitionKeyDefinition)))
```
Comment (neildsh) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Pagination/InMemoryContainer.cs:518]: Nice!!

**Review by neildsh — DISMISSED** [review]

**Comment by neildsh** [comment]
Great description! So just to confirm the change is already covered by existing tests in SubPartitionTests ?

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs [hunk 18]
```diff
@@ -52,7 +52,9 @@ private sealed class InitializationParameters
             public IReadOnlyList<OrderByColumn> OrderByColumns { get; }
             
             public QueryExecutionOptions QueryPaginationOptions { get; }
-            
+
+            public ContainerQueryProperties ContainerQueryProperties { get; }
```
[the remark this answers is not in the document]
Comment (adityasa) [reply Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs:56]: ContainerQueryProperties is a struct and not a class.

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs [hunk 19]
```diff
@@ -71,6 +74,7 @@ private sealed class InitializationParameters
                 this.OrderByColumns = orderByColumns ?? throw new ArgumentNullException(nameof(orderByColumns));
                 this.QueryPaginationOptions = queryPaginationOptions ?? throw new ArgumentNullException(nameof(queryPaginationOptions));
                 this.MaxConcurrency = maxConcurrency;
+                this.ContainerQueryProperties = containerQueryProperties;
```
[the remark this answers is not in the document]
Comment (adityasa) [reply Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs:77]: ContainerQueryProperties is a struct and not a class.

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs (hunk 13 above)
Comment (adityasa) [reply Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs:1837]: ContainerQueryProperties is a struct and not a class.

**Comment by adityasa** [comment]
Correct. I confirmed that the test fails with same same symptoms as reported by cx without the call to HierarchicalPartitionutils from OrderBy Inner Enumerator.

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs (hunk 6 above)
Comment (adityasa) [reply Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs:343]: ContainerQueryProperties is a struct.

**Review by neildsh — APPROVED** [review]

**Review by kirankumarkolli — APPROVED** [review]

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByQueryPartitionRangePageAsyncEnumerator.cs [hunk 20]
```diff
@@ -117,6 +122,7 @@ private sealed class InnerEnumerator : PartitionRangePageAsyncEnumerator<OrderBy
                 : base(feedRangeState)
             {
                 this.queryDataSource = queryDataSource ?? throw new ArgumentNullException(nameof(queryDataSource));
+                this.containerQueryProperties = containerQueryProperties;
```
Comment (Pilchie) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByQueryPartitionRangePageAsyncEnumerator.cs:125]: Nit: How do you decide which arguments you throw for, and which you don't?

**Review by Pilchie — APPROVED** [review]

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByQueryPartitionRangePageAsyncEnumerator.cs (hunk 20 above)
Comment (adityasa) [reply Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByQueryPartitionRangePageAsyncEnumerator.cs:125]: Good point (Neil brought it up too and I ack it's confusing).
ContainerQueryProperties is a struct, so there's no null check needed here.

**Pull request merged**